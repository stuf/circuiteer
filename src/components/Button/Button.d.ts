import { FunctionComponent, ReactNode } from 'react';

interface Props {
  label?: string | ReactNode;
  className?: string;
  size?: 'sm' | 'md' | 'lg';
  type?: 'primary' | 'secondary';
  icon?: string;
}

export const Button: FunctionComponent<Props>;
