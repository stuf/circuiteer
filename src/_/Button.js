import cx from 'classnames';

export default function Button(props) {
  const { children, className, icon: Icon, ...rest } = props;

  const cns = [
    'px-4 py-2 bg-black text-white rounded-md relative',
    'hover:bg-purple-600 hover:shadow-md',
    'focus:bg-purple-500',
    Icon && 'pl-10',
  ];

  return (
    <button className={cx(...cns)} {...rest}>
      {Icon && (
        <span className="absolute inset-y-0 left-2 inline-flex items-center">
          <Icon className="w-5 h-5" />
        </span>
      )}

      {children}
    </button>
  );
}
