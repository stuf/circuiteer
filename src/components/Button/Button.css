.button {
  @apply relative;
  @apply px-4 py-2;
  @apply bg-black text-white;
  @apply rounded;
  @apply hover:bg-purple-600 hover:shadow-md;
  @apply focus:bg-purple-500;

  @apply ring-2 ring-purple-600;
  @apply border-2 border-pink-500;
  @apply ring-offset-2 ring-offset-white;
}

.button.button--with-icon {
  @apply pl-10;
}

.button--with-icon .__icon-wrapper {
  @apply absolute inset-y-0 left-2 inline-flex items-center;
}

.button--with-icon .__icon {
  @apply w-5 h-5;
}
