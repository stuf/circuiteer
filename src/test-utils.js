import { render as renderRtl } from '@testing-library/react';
import { format } from 'util';
import { Provider } from 'react-redux';
import { preloadedState } from './core/preloaded';
import * as reducers from './state';
import { configureStore } from '@reduxjs/toolkit';

import './test/index';

//

export function render(vdom, { initialState, store, ...renderOptions } = {}) {
  const rootStore =
    store ||
    configureStore({
      preloadedState: initialState || preloadedState,
      reducer: reducers,
    });

  function Wrapper({ children }) {
    return <Provider store={rootStore}>{children}</Provider>;
  }

  return renderRtl(vdom, { wrapper: Wrapper, ...renderOptions });
}

//

export function suppressPropTypeWarnings() {
  const err = console.error.bind(console.error);

  beforeAll(() => {
    console.error = (msg, ...rest) => {
      const errorMsg = format(msg, ...rest);
      if (!errorMsg.toString().includes('Warning: Failed prop type')) {
        err(errorMsg);
      }
    };
  });

  afterAll(() => {
    console.error = err;
  });
}
