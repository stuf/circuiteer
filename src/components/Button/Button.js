import * as P from 'prop-types';
import cx from 'classnames';

import { Icon } from 'components';
import './Button.css';

export function Button(props) {
  const { children, className, label, size, type, icon, ...rest } = props;

  const cns = [
    'button',
    icon && 'button--with-icon',
    type && `button--${type}`,
    size && `button--${size}`,
  ];

  return (
    <button className={cx(...cns)} {...rest}>
      {icon && (
        <span className="__icon-wrapper">
          <Icon className="__icon" name={icon} />
        </span>
      )}

      {label || children}
    </button>
  );
}

Button.propTypes = {
  className: P.string,
  icon: P.any,
  label: P.oneOfType([P.string, P.node, P.elementType]),
  type: P.oneOf(['primary', 'secondary']),
  size: P.oneOf(['sm', 'md', 'lg']),
};
