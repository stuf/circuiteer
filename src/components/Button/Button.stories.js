import { Button } from './Button';

export default {
  title: 'UI/Button',
  component: Button,
};

const Template = args => <Button {...args} />;

export const Basic = Template.bind({});
Basic.args = {
  children: 'text',
};
