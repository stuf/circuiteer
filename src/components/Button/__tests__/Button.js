import {
  render,
  createEvent,
  fireEvent,
  logDOM,
  logRoles,
} from '@testing-library/react';
import { Button } from '../Button';
import { suppressPropTypeWarnings } from 'test-utils';

suppressPropTypeWarnings();

test('renders with sensible defaults', () => {
  render(<Button />);
});

it('should create a label text to the button', async () => {
  const { findByLabelText } = render(<Button label="label">children</Button>);

  await findByLabelText('label');
});

it('should not throw when clicking without any given onClick handler', async () => {
  const { findByRole } = render(<Button />);
  const a = await findByRole('button');

  const ev = createEvent.click(a);
  expect(() => fireEvent(a, ev)).not.toThrow();
});

it('should call the given onClick handler when clicked', async () => {
  const fn = jest.fn();
  const { findByRole } = render(<Button onClick={fn} />);

  const el = await findByRole('button');

  fireEvent(el, createEvent.click(el));

  expect(fn).toHaveBeenCalled();
});

it('supports different sizes, types and icons', async () => {
  const { container } = render(
    <>
      <Button icon="settings" label="Settings" />
      <Button icon="settings">Settings</Button>
      <Button icon="settings" label="Settings"></Button>
    </>,
  );

  expect(container).toMatchSnapshot();
});
