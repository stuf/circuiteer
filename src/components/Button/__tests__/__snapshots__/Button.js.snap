// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`supports different sizes, types and icons 1`] = `
<div>
  <button
    aria-label="Settings"
    class="button button--with-icon"
    type="button"
  >
    <span
      aria-hidden="true"
      class="__icon-wrapper"
    >
      <span
        class="icon icon--uiflex items-center justify-center"
        style="width: 20px; height: 20px;"
      >
        <span
          class="material-icons material-icons-outlined"
          style="font-size: 20px;"
        >
          settings
        </span>
      </span>
    </span>
  </button>
  <button
    class="button button--with-icon"
    type="button"
  >
    <span
      aria-hidden="true"
      class="__icon-wrapper"
    >
      <span
        class="icon icon--uiflex items-center justify-center"
        style="width: 20px; height: 20px;"
      >
        <span
          class="material-icons material-icons-outlined"
          style="font-size: 20px;"
        >
          settings
        </span>
      </span>
    </span>
    Settings
  </button>
  <button
    aria-label="Settings"
    class="button button--with-icon"
    type="button"
  >
    <span
      aria-hidden="true"
      class="__icon-wrapper"
    >
      <span
        class="icon icon--uiflex items-center justify-center"
        style="width: 20px; height: 20px;"
      >
        <span
          class="material-icons material-icons-outlined"
          style="font-size: 20px;"
        >
          settings
        </span>
      </span>
    </span>
  </button>
</div>
`;
